api/client/service/update_test.go

```diff
@@ -179,3 +179,41 @@ func TestUpdatePortsConflictingFlags(t *testing.T) {
 	err := updatePorts(flags, &portConfigs)
 	assert.Error(t, err, "conflicting port mapping")
 }
+
+func TestUpdatePortsRmWithProtocol(t *testing.T) {
+	flags := newUpdateCommand(nil).Flags()
+	flags.Set("publish-add", "8081:81")
+	flags.Set("publish-add", "8082:82")
+	flags.Set("publish-rm", "80")
+	flags.Set("publish-rm", "81/tcp")
+	flags.Set("publish-rm", "82/udp")
+
+	portConfigs := []swarm.PortConfig{
+		{TargetPort: 80, PublishedPort: 8080, Protocol: swarm.PortConfigProtocolTCP},
+	}
+
+	err := updatePorts(flags, &portConfigs)
+	assert.Equal(t, err, nil)
+	assert.Equal(t, len(portConfigs), 1)
+	assert.Equal(t, portConfigs[0].TargetPort, uint32(82))
+}
+
+func TestValidatePort(t *testing.T) {
+	validPorts := []string{"80/tcp", "80", "80/udp"}
+	invalidPorts := map[string]string{
+		"9999999": "out of range",
+		"80/xyz":  "invalid protocol",
+		"tcp":     "invalid syntax",
+		"udp":     "invalid syntax",
+		"":        "invalid protocol",
+	}
```

Comment:
I think we're missing an important test case here for `80:80`. We should make sure to give a helpful error in that case.